File: RailgunVB/Core/Utilities/CommandUtils.vb
Imports AudioChord
Imports Discord
Imports Microsoft.Extensions.DependencyInjection
Imports MongoDB.Bson
Imports TreeDiagram
Imports TreeDiagram.Models.Server
Imports TreeDiagram.Models.User

Namespace Core.Utilities
    
    Public Class CommandUtils
    
        Private ReadOnly _services As IServiceProvider
        Private ReadOnly _musicService As MusicService

        Public Sub New(services As IServiceProvider)
            _musicService = services.GetService(Of MusicService)
            _services = services
        End Sub
        
        Public Async Function GetUsernameOrMentionAsync(user As IGuildUser) As Task(Of String)
            Dim sMention As ServerMention
            Dim uMention As UserMention
            
            Using scope As IServiceScope = _services.CreateScope()
                Dim db As TreeDiagramContext = scope.ServiceProvider.GetService(Of TreeDiagramContext)
                sMention = await db.ServerMentions.GetAsync(user.GuildId)
                uMention = await db.UserMentions.GetAsync(user.Id)
            End Using
            
            If (sMention IsNot Nothing AndAlso sMention.DisableMentions) OrElse 
               (uMention IsNot Nothing AndAlso uMention.DisableMentions)
                Return user.Username
            Else 
                Return user.Mention
            End If
        End Function
        
        Public Async Function GetPlaylistAsync(data As ServerMusic) As Task(Of Playlist)
            If Not (data.PlaylistId = ObjectId.Empty) Then Return Await _musicService.Playlist.GetPlaylistAsync(
                data.PlaylistId)

            Dim playlist As New Playlist
            
            data.PlaylistId = playlist.Id
            
            Await _musicService.Playlist.UpdateAsync(playlist)
            
            Return playlist
        End Function
        
        Public Async Function CheckIfSelfIsHigherRole(guild As IGuild, user As IGuildUser) As Task(Of Boolean)
            Dim selfRolePosition = 0
            Dim userRolePosition = 0
            Dim self As IGuildUser = Await guild.GetCurrentUserAsync()
            
            For Each roleId As ULong In self.RoleIds
                Dim role As IRole = Guild.GetRole(roleId)
                If role.Permissions.BanMembers AndAlso 
                   role.Position > selfRolePosition Then selfRolePosition = role.Position
            Next
            
            For Each roleId As ULong In user.RoleIds
                Dim role As IRole = Guild.GetRole(roleId)
                If role.Position > userRolePosition Then userRolePosition = role.Position
            Next
            
            If selfRolePosition > userRolePosition Then Return True
            Return False
        End Function
        
    End Class
    
End NameSpace

File: Railgun-Old/Core/Utilities/CommandUtils.cs
using System;
using System.IO;
using System.Text;
using System.Threading.Tasks;
using AudioChord;
using Discord;
using Microsoft.Extensions.DependencyInjection;
using MongoDB.Bson;
using TreeDiagram;
using TreeDiagram.Models.Server;
using TreeDiagram.Models.User;

namespace Railgun.Core.Utilities
{
	public class CommandUtils
	{
		private readonly IServiceProvider _services;
		private readonly MusicService _musicService;

		public CommandUtils(IServiceProvider services)
		{
			_services = services;

			_musicService = _services.GetService<MusicService>();
		}

		public string GetUsernameOrMention(IGuildUser user)
		{
			ServerMention sMention;
			UserMention uMention;

			using (var scope = _services.CreateScope()) {
				var db = scope.ServiceProvider.GetService<TreeDiagramContext>();

				sMention = db.ServerMentions.GetData(user.GuildId);
				uMention = db.UserMentions.GetData(user.Id);
			}

			if ((sMention != null && sMention.DisableMentions) || (uMention != null && uMention.DisableMentions))
				return user.Username;
			return user.Mention;
		}

		public async Task<Playlist> GetPlaylistAsync(ServerMusic data)
		{
			if (data.PlaylistId != ObjectId.Empty)
				return await _musicService.Playlist.GetPlaylistAsync(data.PlaylistId);

			var playlist = new Playlist();

			data.PlaylistId = playlist.Id;

			await _musicService.Playlist.UpdateAsync(playlist);

			return playlist;
		}

		public async Task<bool> CheckIfSelfIsHigherRole(IGuild guild, IGuildUser user)
		{
			var selfRolePosition = 0;
			var userRolePosition = 0;
			var self = await guild.GetCurrentUserAsync();

			foreach (var roleId in self.RoleIds) {
				var role = guild.GetRole(roleId);

				if (role.Permissions.BanMembers && role.Position > selfRolePosition)
					selfRolePosition = role.Position;
			}

			foreach (var roleId in user.RoleIds) {
				var role = guild.GetRole(roleId);

				if (role.Position > userRolePosition) userRolePosition = role.Position;
			}

			if (selfRolePosition > userRolePosition) return true;
			else return false;
		}

		public static async Task SendStringAsFileAsync(ITextChannel tc, string filename, string output, string msgText = null, bool includeGuildName = true)
		{
			var outputStream = new MemoryStream(Encoding.UTF8.GetBytes(output));
			var outputFilename = (includeGuildName ? $"{tc.Guild.Name}-" : "") + filename;

			await tc.SendFileAsync(outputStream, outputFilename, msgText);
		}
	}
}